### PR TITLE
Import BGEN v1.2

### PR DESCRIPTION
**Pull request by jigold**

### CODE REVIEW

**Comment by jigold** [comment]
- Only support diploid and unphased genotype probabilities

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+      fatal("Hail does not support phased genotypes.")
+
+    val nBitsPerProb = bytes(nSamples + 9) & 0xff
+    assert(nBitsPerProb >= 1 && nBitsPerProb <= 32, s"Value for nBits must be between 1 and 32 inclusive. Found $nBitsPerProb.")
+
+    val probabilityIterator = new BgenProbabilityIterator(bytes, nBitsPerProb, nSamples + 10)
+
+    val nGenotypes = triangle(nAlleles)
+    val sampleProbs = Array.ofDim[Long](nGenotypes)
+    val totalProbInt = ~0L >>> (64 - nBitsPerProb)
```
Comment (cseed) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: More traditional to write `(1L << nBitsPerProb) - 1`.  That gets you the low nBitsPerProb bits set.

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    if (isPhased)
+      fatal("Hail does not support phased genotypes.")
+
+    val nBitsPerProb = bytes(nSamples + 9) & 0xff
+    assert(nBitsPerProb >= 1 && nBitsPerProb <= 32, s"Value for nBits must be between 1 and 32 inclusive. Found $nBitsPerProb.")
+
+    val probabilityIterator = new BgenProbabilityIterator(bytes, nBitsPerProb, nSamples + 10)
+
+    val nGenotypes = triangle(nAlleles)
+    val sampleProbs = Array.ofDim[Long](nGenotypes)
```
Comment (cseed) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: We normally write `new Array[Long](nGenotypes)`.

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    val maxPloidy = bytes(7) & 0xff
+
+    if (minPloidy != 2 || maxPloidy != 2)
+      fatal(s"Hail only supports diploid genotypes. Found min ploidy equals `$minPloidy' and max ploidy equals `$maxPloidy'.")
+
+    val samplePloidy = Array.ofDim[Int](nSamples)
+    var i = 0
+    while (i < nSamples) {
+      val b = bytes(i + 8)
+      val isMissing = ((b & 0xffL) >>> 7) == 1
```
Comment (cseed) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: Clearer to write `(b & 0x80) != 0`.

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+    new Iterable[Genotype] {
+      def iterator = new Iterator[Genotype] {
+        var sampleIndex = 0
+
+        def hasNext: Boolean = sampleIndex < nSamples
+
+        def next(): Genotype = {
+          var i = 0
+          var sumProbInt = 0L
```
Comment (cseed) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: `sumProbInt` is a `Long`?

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala (hunk 1 above)
Comment (cseed) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: `totalProbInt` is a `Long`?

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+    if (minPloidy != 2 || maxPloidy != 2)
+      fatal(s"Hail only supports diploid genotypes. Found min ploidy equals `$minPloidy' and max ploidy equals `$maxPloidy'.")
+
+    val samplePloidy = Array.ofDim[Int](nSamples)
+    var i = 0
+    while (i < nSamples) {
+      val b = bytes(i + 8)
+      val isMissing = ((b & 0xffL) >>> 7) == 1
+      val ploidy = if (isMissing) -1 else b & 0x3f
+      assert(ploidy >= -1 && ploidy <= 63, s"Ploidy value must be in range [0, 63]. Found $ploidy.")
```
Comment (cseed) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: ploidy has to equal 2 by min/max above, right?  I think you can simplify this.

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+            val p = probabilityIterator.next()
+            sampleProbs(i) = p
+            i += 1
+            sumProbInt += p
+          }
+
+          assert(sumProbInt <= totalProbInt, "Sum of probabilities is greater than 1.")
+          sampleProbs(i) = totalProbInt - sumProbInt
+
+          val gt = if (samplePloidy(sampleIndex) == -1)
```
Comment (cseed) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: ploidy will always be 2, always false.

File: src/main/scala/is/hail/io/bgen/BgenProbabilityIterator.scala [hunk 7]
```diff
@@ -0,0 +1,25 @@
+package is.hail.io.bgen
+
+class BgenProbabilityIterator(input: Array[Byte], nBitsPerProb: Int, start: Int = 0) extends Iterator[Long] {
+  val nBytes = input.length
+  val bitMask = ~0L >>> (64 - nBitsPerProb)
```
Comment (cseed) [inline src/main/scala/is/hail/io/bgen/BgenProbabilityIterator.scala]: `(1 << nBitsPerProb) - 1`

File: src/main/scala/is/hail/io/bgen/BgenProbabilityIterator.scala [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+  var data = 0L
+  var dataSize = 0
+
+  override def next(): Long = {
+    while (dataSize < nBitsPerProb && byteIndex < nBytes) {
+      data |= ((input(byteIndex) & 0xffL) << dataSize)
+      byteIndex += 1
+      dataSize += 8
+    }
+
```
Comment (cseed) [inline src/main/scala/is/hail/io/bgen/BgenProbabilityIterator.scala]: might assert dataSize >= nBitsPerProb.

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+    assert(nBitsPerProb >= 1 && nBitsPerProb <= 32, s"Value for nBits must be between 1 and 32 inclusive. Found $nBitsPerProb.")
+
+    val probabilityIterator = new BgenProbabilityIterator(bytes, nBitsPerProb, nSamples + 10)
+
+    val nGenotypes = triangle(nAlleles)
+    val sampleProbs = Array.ofDim[Long](nGenotypes)
+    val totalProbInt = ~0L >>> (64 - nBitsPerProb)
+    val noCall: Genotype = new DosageGenotype(-1, null)
+
+    val nExpectedBytesProbs = math.ceil((nSamples * (nGenotypes - 1) * nBitsPerProb).toDouble / 8).toInt
```
Comment (cseed) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: You don't need Double math.  Write `nSamples *  (nGenotypes - 1) * nBitsPerProb + 7) / 8`.  `... + 7) / 8` has the effect of rounding up.

File: src/test/scala/is/hail/io/LoadBgenSuite.scala [hunk 10]
```diff
@@ -136,4 +151,71 @@ class LoadBgenSuite extends SparkSuite {
...
+      v <- Gen.buildableOf[Array, Double](Gen.choose(0d, 1d)).resize(10);
+      nBitsPerProb <- Gen.choose(1, 32)) yield (v, nBitsPerProb)
+
+    property("bgen probability iterator gives correct result") =
+      forAll(Gen.buildableOf[Array, Double](Gen.choose(0d, 1d)).resize(10)) { v =>
+        val probs = v.map(_ / v.sum)
+
+        (1 to 32).forall { nBitsPerProb =>
+
+          val probInts = probs.map(_ * (math.pow(2, nBitsPerProb) - 1))
```
Comment (cseed) [inline src/test/scala/is/hail/io/LoadBgenSuite.scala]: No pow'ing powers of 2!  `(1 << nBitsPerProb) - 1`.

File: src/test/scala/is/hail/io/LoadBgenSuite.scala [hunk 11]
```diff
@@ -136,4 +151,71 @@ class LoadBgenSuite extends SparkSuite {
...
+          val floorInts = probInts.map(i => (i, math.floor(i)))
+          val totalFractional = floorInts.map(f => f._1 - f._2).sum.toInt
+          val input = floorInts.sortBy(i => -(i._1 - i._2)).zipWithIndex.map { case (r, i) =>
+            if (i < totalFractional)
+              (i, math.ceil(r._1))
+            else
+              (i, math.floor(r._1))
+          }.map(_._2.toInt)
```
Comment (cseed) [inline src/test/scala/is/hail/io/LoadBgenSuite.scala]: I find this i a bit hard to read.  You just want to partition the total probability, right?  Isn't it easier to write `for (nBitsPerProb <- 1 to 32; input <- Gen.partition((1 << nBitsPerProb) - 1, 3) { ...` or something like that?

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+
+  def setExpectedNumAlleles(n: Int) {
+    this.expectedNumAlleles = n
+  }
+
+  override def getValue: Iterable[Genotype] = {
+    require(input != null, "called getValue before serialized value was set")
+
+    val bytes = if (compressed) decompress(input, expectedDataSize) else input
+    val nRow = (bytes(0) & 0xff) | ((bytes(1) & 0xff) << 8) | ((bytes(2) & 0xff) << 16) | ((bytes(3) & 0xff) << 24)
```
Comment (cseed) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: You might write some utility functions for grabbing out little endian shorts, ints from the byte array.

File: python/hail/context.py [hunk 13]
```diff
@@ -147,25 +147,27 @@ def import_bgen(self, path, tolerance=0.2, sample_file=None, npartitions=None):
...
+        see `here <http://www.well.ox.ac.uk/~gav/bgen_format/bgen_format.html>`__. Note that only v1.1 and v1.2 BGEN files
+        are supported at this time. For v1.2 BGEN files, only **unphased** and **diploid** genotype probabilities are allowed and the
+        genotype probability blocks must be either compressed with zlib or uncompressed.
 
         Before importing, ensure that:
-        - Files reside in Hadoop file system.
-        - The sample file has the same number of samples as the BGEN file.
-        - No duplicate sample IDs are present.
+
+          - Files reside in Hadoop file system.
```
Comment (cseed) [inline python/hail/context.py]: You might kill this, users are using a variety of filesystems.

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala (hunk 6 above)
Comment (jigold) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: No -- ploidy can be equal to -1 (I chose this as the missing value -- now that I think about it, I can just test if the ploidy is negative because missing is defined as the most significant bit set to 1). They still emit probabilities for No Calls by having two probabilities equal to 0 in the diploid/biallelic case. If the ploidy wasn't set to missing, you'd interpret the probabilities as Array(0.0, 0.0, 1.0) because the third probability is deduced from the others.

File: src/test/scala/is/hail/io/LoadBgenSuite.scala (hunk 11 above)
Comment (jigold) [inline src/test/scala/is/hail/io/LoadBgenSuite.scala]: I'll rewrite it as you suggest for this test. In the exportBGEN branch, the BGEN writer needs to do this to follow the rounding rules in the spec.

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala (hunk 6 above)
Comment (cseed) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: Ah, OK, makes sense.

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+    assert(nRow == nSamples, "row nSamples is not equal to header nSamples")
+
+    val nAlleles = (bytes(4) & 0xff) | ((bytes(5) & 0xff) << 8)
+    assert(nAlleles == expectedNumAlleles, s"Value for `nAlleles' in genotype probability data storage is not equal to value in variant identifying data. Expected $expectedNumAlleles but found $nAlleles.")
+
+    val minPloidy = bytes(6) & 0xff
+    val maxPloidy = bytes(7) & 0xff
+
+    if (minPloidy != 2 || maxPloidy != 2)
+      fatal(s"Hail only supports diploid genotypes. Found min ploidy equals `$minPloidy' and max ploidy equals `$maxPloidy'.")
```
Comment (tpoterba) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: do we know that the sex chromosomes are encoded as ploidy 2 for males? This could be a problem

Comment (jigold) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: According to Gavin's wiki page, all genotypes will be diploid. 
https://bitbucket.org/gavinband/bgen/wiki/BGEN_in_the_UK_Biobank

**Comment by jigold** [comment]
@cseed back to you

**Comment by jigold** [comment]
@cseed please hold off on this until I incorporate changes in the export branch. I want to make sure everything is working with the import/export test.

**Comment by jigold** [comment]
@cseed this is good to go now.

File: src/main/scala/is/hail/utils/UInt.scala [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+  def <(right: UInt): Boolean = toLong < right.toLong
+
+  def >(right: UInt): Boolean = toLong > right.toLong
+
+  def ==(right: UInt): Boolean = toLong == right.toLong
+
+  def !=(right: UInt): Boolean = toLong != right.toLong
+
+  def toInt: Int = {
+    assert(i >= 0)
```
Comment (cseed) [inline src/main/scala/is/hail/utils/UInt.scala]: The invariant (i >= 0) should be on the object, not the accessor.  Right now, I can say UInt(-3) + UInt(5) == UInt(2) and it will be true.

File: src/main/scala/is/hail/utils/UInt.scala [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+    override def compare(x: UInt, y: UInt): Int = x.compare(y)
+
+    override def zero: UInt = UInt(0)
+
+    override def one: UInt = UInt(1)
+  }
+}
+
+class UInt private(val i: Int) extends AnyVal {
+
```
Comment (cseed) [inline src/main/scala/is/hail/utils/UInt.scala]: Check i >= 0 here.

File: src/main/scala/is/hail/utils/UInt.scala [hunk 17]
```diff
@@ -0,0 +1,78 @@
+package is.hail.utils
+
+object UInt {
+  def apply(i: Int): UInt = new UInt(i)
+
+  def apply(l: Long): UInt = {
+    assert((l >>> 32) == 0, s"High bits not set to 0. Cannot convert $l to UInt.")
```
Comment (cseed) [inline src/main/scala/is/hail/utils/UInt.scala]: I'd make this message more "semantic": Long value does not fit in UInt.

File: src/main/scala/is/hail/io/bgen/BgenBlockReader.scala [hunk 18]
```diff
@@ -153,3 +77,59 @@ class BgenBlockReader(job: Configuration, split: FileSplit) extends IndexedBinar
...
+      val rsid = bfis.readLengthAndString(2)
+      val chr = bfis.readLengthAndString(2)
+      val position = bfis.readInt()
+
+      val nAlleles = bfis.readShort()
+      assert(nAlleles >= 2, s"Number of alleles must be greater than or equal to 2. Found $nAlleles alleles for variant '$lid'")
+      val nAltAlleles = nAlleles - 1
+
+      val ref = bfis.readLengthAndString(4)
+      val altAlleles = Array.ofDim[String](nAltAlleles)
```
Comment (cseed) [inline src/main/scala/is/hail/io/bgen/BgenBlockReader.scala]: I think we've been using `new Array[String](nAltAllles)` instead of `ofDim`.

File: src/main/scala/is/hail/io/bgen/BgenBlockReader.scala [hunk 19]
```diff
@@ -153,3 +77,59 @@ class BgenBlockReader(job: Configuration, split: FileSplit) extends IndexedBinar
...
+      }
+
+      val variant = Variant(recodedChr, position, ref, altAlleles)
+
+      val dataSize = bfis.readInt()
+      val uncompressedSize = if (bState.compressed) bfis.readInt() else dataSize
+
+      val bytesInput =
+        if (bState.compressed)
+          bfis.readBytes(dataSize - 4)
```
Comment (cseed) [inline src/main/scala/is/hail/io/bgen/BgenBlockReader.scala]: The compressed size includes uncompressedSize?  You might combine these two ifs:

```
  val (uncompressedSize, bytesInput) = ...
```

File: src/main/scala/is/hail/io/bgen/BgenLoader.scala [hunk 20]
```diff
@@ -101,21 +108,27 @@ object BgenLoader {
       for (i <- 1 to bState.nVariants) {
         reader.seek(position)
 
-        val nRow = reader.readInt()
+        if (bState.version == 1)
+          reader.readInt()
```
Comment (cseed) [inline src/main/scala/is/hail/io/bgen/BgenLoader.scala]: Might comment this value since you're not assigning it to any name.

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala [hunk 21]
```diff
@@ -0,0 +1,176 @@
...
+    val nBitsPerProb = reader.read()
+    assert(nBitsPerProb >= 1 && nBitsPerProb <= 32, s"Value for nBits must be between 1 and 32 inclusive. Found $nBitsPerProb.")
+
+    val nGenotypes = triangle(nAlleles)
+    val nExpectedBytesProbs = (nSamples * (nGenotypes - 1) * nBitsPerProb + 7) / 8
+    assert(reader.length == nExpectedBytesProbs + nSamples + 10, s"Number of uncompressed bytes `${ reader.length }' does not match the expected size `$nExpectedBytesProbs'.")
+
+    val probabilityIterator = new BgenProbabilityIterator(reader, nBitsPerProb)
+
+    val sampleProbs = Array.fill[UInt](nGenotypes)(UInt(0))
```
Comment (cseed) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: new

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala [hunk 22]
```diff
@@ -0,0 +1,176 @@
...
+
+    val totalProbInt = UInt((1 << nBitsPerProb) - 1)
```
Comment (cseed) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: UInt(x) or x.toUInt?  Should we prefer one or the other?  Latter is more like the builtin types.

File: src/main/scala/is/hail/utils/package.scala [hunk 23]
```diff
@@ -455,4 +456,15 @@ package object utils extends Logging
     }
     s
   }
+
+  def decompress(input: Array[Byte], size: Int): Array[Byte] = {
+    val expansion = Array.ofDim[Byte](size)
```
Comment (cseed) [inline src/main/scala/is/hail/utils/package.scala]: new

File: src/main/scala/is/hail/utils/richUtils/RichLong.scala [hunk 24]
```diff
@@ -0,0 +1,10 @@
+package is.hail.utils.richUtils
+
+import is.hail.utils.UInt
+
+class RichLong(val l: Long) extends AnyVal {
+  def toUInt: UInt = {
+    assert((l >>> 32) == 0, s"High bits not set to 0. Cannot convert ${ l.toBinaryString } to UInt.")
```
Comment (cseed) [inline src/main/scala/is/hail/utils/richUtils/RichLong.scala]: Duplicate code.  You have a long constructor (UInt.apply), use that.

File: src/main/scala/is/hail/utils/UInt.scala (hunk 16 above)
Comment (jigold) [inline src/main/scala/is/hail/utils/UInt.scala]: If I assert i >= 0, then the max value is 2^31 - 1. We need 2^32 which is -1.

File: src/main/scala/is/hail/utils/UInt.scala (hunk 15 above)
Comment (jigold) [inline src/main/scala/is/hail/utils/UInt.scala]: I tried this and it throws an assertion error.

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala (hunk 21 above)
Comment (jigold) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: If I say `new Array[UInt](nGenotypes)`, the array is initialized with null values which is not what I wanted.

Comment (jigold) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: Although, I think it's fine in this case because every element is filled in with the for loop.

File: src/main/scala/is/hail/io/bgen/BgenBlockReader.scala (hunk 19 above)
Comment (jigold) [inline src/main/scala/is/hail/io/bgen/BgenBlockReader.scala]: No, the compressed size is the first Int and the uncompressed size is the second Int (but only present if the block is actually compressed). So you have `dataSize - 4` to account for the missing Int in the case where the genotype block hasn't been compressed.

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala (hunk 21 above)
Comment (cseed) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: UInt is any AnyVal so it is initialized to the default value of its singleton type, Int, which is 0.

File: src/main/scala/is/hail/utils/UInt.scala (hunk 15 above)
Comment (cseed) [inline src/main/scala/is/hail/utils/UInt.scala]: Ah, sorry, clearly require(i >= 0) in the class is wrong, since we can have unsigned numbers with the high bit set.  It should be tested in the UInt.apply(Int).  Right now it looks like if you construct a UInt with a negative number UInt(-2), it will work.  Am I wrong?

Comment (jigold) [inline src/main/scala/is/hail/utils/UInt.scala]: I changed the behavior to no longer accept a negative integer in the constructor.

**Comment by jigold** [comment]
@cseed back to you.
